test/Test/Database/LSMTree/Normal/StateMachine.hs

```diff
@@ -0,0 +1,696 @@
+           , findBlobRefsVars
+           ) of
+        ([], _ )      -> []
+        (_ , [])      -> []
+        (vars, vars') -> withVars' (QC.elements vars) (QC.elements vars')
+    ]
+  where
+    findBlobRefsVars :: [Var (Either Model.Err [SUT.BlobRef blob])]
+    findBlobRefsVars = fmap fromLookupResults vars1 ++ fmap fromRangeLookupResults vars2
+      where
+        vars1 = findVars (Proxy @(Either Model.Err [SUT.LookupResult k v (SUT.BlobRef blob)]))
+        vars2 = findVars (Proxy @(Either Model.Err [SUT.RangeLookupResult k v (SUT.BlobRef blob)]))
+
+        fromLookupResults ::
+             Var (Either Model.Err [SUT.LookupResult k v (SUT.BlobRef blob)])
+          -> Var (Either Model.Err [SUT.BlobRef blob])
+        fromLookupResults = mapGVar (\op -> OpRight `OpComp` OpLookupResults `OpComp` OpFromRight `OpComp` op)
+
+        fromRangeLookupResults ::
+             Var (Either Model.Err [SUT.RangeLookupResult k v (SUT.BlobRef blob)])
+          -> Var (Either Model.Err [SUT.BlobRef blob])
+        fromRangeLookupResults = mapGVar (\op -> OpRight `OpComp` OpRangeLookupResults `OpComp` OpFromRight `OpComp` op)
+
+    withoutVars :: [Gen (Any (LockstepAction ModelState))]
+    withoutVars = [
+          Some . New @k @v @blob <$> genTableConfig
+        ]
+
+    withVars ::
+         Gen (Var (Either Model.Err (SUT.TableHandle IO k v blob)))
+      -> [Gen (Any (LockstepAction ModelState))]
+    withVars genVar = [
+          fmap Some $ Close <$> (fromRight <$> genVar)
+        , fmap Some $ Lookups <$> genLookupKeys <*> (fromRight <$> genVar)
+        , fmap Some $ RangeLookup <$> genRange <*> (fromRight <$> genVar)
+        , fmap Some $ Updates <$> genUpdates <*> (fromRight <$> genVar)
+        , fmap Some $ Inserts <$> genInserts <*> (fromRight <$> genVar)
+        , fmap Some $ Deletes <$> genDeletes <*> (fromRight <$> genVar)
+        , fmap Some $ Snapshot <$> genSnapshotName <*> (fromRight <$> genVar)
+        , fmap Some $ Open @k @v @blob <$> genSnapshotName
+        , fmap Some $ Duplicate <$> (fromRight <$> genVar)
+        ]
+
+    withVars' ::
+         Gen (Var (Either Model.Err (SUT.TableHandle IO k v blob)))
+      -> Gen (Var (Either Model.Err [SUT.BlobRef blob]))
+      -> [Gen (Any (LockstepAction ModelState))]
+    withVars' genTableHandleVar genBlobRefsVar = [
+          fmap Some $ RetrieveBlobs <$> (fromRight <$> genTableHandleVar) <*> (fromRight <$> genBlobRefsVar)
+        ]
+
+    fromRight ::
+         Var (Either Model.Err a)
+      -> Var a
+    fromRight = mapGVar (\op -> OpFromRight `OpComp` op)
+
+    -- TODO: improve, more types of configs
+    genTableConfig :: Gen SUT.TableConfig
+    genTableConfig = pure $ SUT.TableConfig {
+          SUT.tcMaxBufferMemory = 2 * 1024 * 1024
+        , SUT.tcMaxBloomFilterMemory = 2 * 1024 * 1024
+        , SUT.tcBitPrecision = ()
+        }
+
+    -- TODO: improve
+    genLookupKeys :: Gen [k]
+    genLookupKeys = QC.arbitrary
```

Comment:
So what I did for more interesting generation of keys to insert/delete/lookup was to follow the advice from Edsko's blog post about generating (& shrinking) with dependencies.

The lsm handles are obviously variables, but the trick is to notice that the choice of key to insert/delete/lookup is also a dependency and can be turned into a variable.

Here's a snippet from a different (otherwise simpler) QLS test:
```
    AInsert :: ModelVar Model (LSM RealWorld)
            -> Either (ModelVar Model Key) Key -- to refer to a prior key
            -> Value
            -> Action (Lockstep Model) (Key)

    ADelete :: ModelVar Model (LSM RealWorld)
            -> Either (ModelVar Model Key) Key
            -> Action (Lockstep Model) ()
   
    ALookup :: ModelVar Model (LSM RealWorld)
            -> Either (ModelVar Model Key) Key
            -> Action (Lockstep Model) (Maybe Value)
```
The point is that insert will return the key that was selected, and so that becomes a variable that can be reused in later commands. Then each of these commands accepts as an input key _either_ a normal key value or a variable that references a key that was used earlier (by insert). Then the generation of commands can select a mix of ordinary keys or dependent keys.

For example:
```
        frequency $
            -- inserts of potentially fresh keys
          [ (3, fmap Some $
                  AInsert <$> elements vars
                          <*> fmap Right arbitrarySizedNatural -- key
                          <*> arbitrarySizedNatural)           -- value
          ]
            -- inserts of the same keys as used earlier
       ++ [ (1, fmap Some $
                  AInsert <$> elements vars
                          <*> fmap Left (elements kvars) -- key var
                          <*> arbitrarySizedNatural)    -- value
          | let kvars = findVars (Proxy :: Proxy Key)
          , not (null kvars)
          ]
```
and similarly for deletes and lookups.

And the shrinking can also do something with them. What I found worked quite well was to try shrinking variables to non-variables:
```
  shrinkWithVars findVars _model (AInsert var (Right k) v) =
    [ Some $ AInsert var (Right k') v' | (k', v') <- shrink (k, v) ]

  shrinkWithVars findVars _model (AInsert var (Left _kv) v) =
    [ Some $ AInsert var (Right k) v | k <- shrink 100 ]
```